Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,239 @@
+                        }
+                    }
+
+                case .serviceThrew(let service, let error):
+                    // One of the servers threw an error. We have to cancel everything else now.
+                    self.logger.error(
+                        "Service threw error. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                            self.configuration.loggingConfiguration.errorKey: "\(error)",
+                        ]
+                    )
+                    group.cancelAll()
+
+                    throw error
+
+                case .signalCaught(let unixSignal):
+                    // We got a signal. Let's initiate graceful shutdown in reverse order than we started the
+                    // services. This allows the users to declare a hierarchy with the order they passed
+                    // the services.
+                    self.logger.info(
+                        "Signal caught",
+                        metadata: [
+                            self.configuration.loggingConfiguration.signalKey: "\(unixSignal)",
+                        ]
+                    )
+
+                    for service in self.services.reversed() {
+                        self.logger.debug(
+                            "Shutting down service",
+                            metadata: [
+                                self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                            ]
+                        )
+
+                        do {
+                            try await service.shutdownGracefully()
+                        } catch {
+                            // If we fail to gracefully shutdown a service there is really not much left
+                            // we can do besides cancelling everything.
+                            self.logger.error(
+                                "Service threw error while gracefully shutting down. Cancelling all other services now",
+                                metadata: [
+                                    self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                                    self.configuration.loggingConfiguration.errorKey: "\(error)",
+                                ]
+                            )
+
+                            group.cancelAll()
+                            throw error
+                        }
+                    }
+
+                case .signalSequenceFinished:
+                    // This can happen when we are either cancelling everything or
+                    // when the user did not specify any shutdown signals. We just have to tolerate
+                    // this.
+                    continue
+
+                case nil:
+                    // This is a bit weird but I guess it can happen
+                    return
```

Comment:
I don't think this can happen. we checked if the group is empty in the while loop.

Reply:
I agree it shouldn't happen :D